--- dropbear/libtomcrypt/src/headers/tomcrypt_dropbear.h
+++ dropbear/libtomcrypt/src/headers/tomcrypt_dropbear.h
@@ -27,7 +27,7 @@
 #define LTC_DES
 #endif
 
-#if DROPBEAR_ENABLE_CTR_MODE
+#if DROPBEAR_ENABLE_CBC_MODE
 #define LTC_CBC_MODE
 #endif
 
--- dropbear/session.h
+++ dropbear/session.h
@@ -77,7 +77,9 @@ struct key_context_directional {
 #endif
 	/* actual keys */
 	union {
+#if DROPBEAR_ENABLE_CBC_MODE
 		symmetric_CBC cbc;
+#endif
 #if DROPBEAR_ENABLE_CTR_MODE
 		symmetric_CTR ctr;
 #endif
