--- dnsmasq/src/dhcp.c
+++ dnsmasq/src/dhcp.c
@@ -152,8 +152,10 @@
   time_t recvtime = now;
 #ifdef HAVE_LINUX_NETWORK
   struct arpreq arp_req;
+#ifndef HAVE_TOMATO
   struct timeval tv;
 #endif
+#endif
   
   union {
     struct cmsghdr align; /* this ensures alignment */
@@ -179,8 +181,10 @@
     return;
     
   #if defined (HAVE_LINUX_NETWORK)
+#ifndef HAVE_TOMATO
   if (ioctl(fd, SIOCGSTAMP, &tv) == 0)
     recvtime = tv.tv_sec;
+#endif
 
   if (msg.msg_controllen >= sizeof(struct cmsghdr))
     for (cmptr = CMSG_FIRSTHDR(&msg); cmptr; cmptr = CMSG_NXTHDR(&msg, cmptr))
