--- dnsmasq/src/dnssec.c
+++ dnsmasq/src/dnssec.c
@@ -899,7 +899,10 @@ int dnssec_validate_ds(time_t now, struc
   
   if (rc == STAT_INSECURE)
     {
-      my_syslog(LOG_WARNING, _("Insecure DS reply received, do upstream DNS servers support DNSSEC?"));
+      if ((strcmp(name, "com") != 0) && (strcmp(name, "COM") != 0))
+        my_syslog(LOG_WARNING, _("Insecure DS reply received from %s"), name);
+      else
+        my_syslog(LOG_WARNING, _("Insecure DS reply received, do upstream DNS servers support DNSSEC?"));
       rc = STAT_BOGUS;
     }
   
