--- openssl-1.1/Configurations/10-main.conf
+++ openssl-1.1/Configurations/10-main.conf
@@ -760,6 +760,18 @@ my %targets = (
         inherit_from     => [ "linux-generic64"],
         perlasm_scheme   => "linux64",
     },
+    "linux-mipsel" => {
+        inherit_from     => [ "linux-mips32" ],
+        CFLAGS           => add(picker(release => "-fomit-frame-pointer \$(EXTRACFLAGS)")),
+        cppflags         => add("-D_XOPEN_SOURCE=500"),
+        lib_cppflags     => add("-DL_ENDIAN"),
+    },
+    "linux-mips" => {
+        inherit_from     => [ "linux-mips32" ],
+        CFLAGS           => add(picker(release => "-fomit-frame-pointer \$(EXTRACFLAGS)")),
+        cppflags         => add("-D_XOPEN_SOURCE=500"),
+        lib_cppflags     => add("-DB_ENDIAN"),
+    },
 
     #### IA-32 targets...
     #### These two targets are a bit aged and are to be used on older Linux
--- openssl-1.1/crypto/build.info
+++ openssl-1.1/crypto/build.info
@@ -10,7 +10,7 @@ EXTRA=  ../ms/uplink-x86.pl ../ms/uplink
         ppccpuid.pl pariscid.pl alphacpuid.pl arm64cpuid.pl armv4cpuid.pl
 
 DEPEND[cversion.o]=buildinf.h
-GENERATE[buildinf.h]=../util/mkbuildinf.pl "$(CC) $(LIB_CFLAGS) $(CPPFLAGS_Q)" "$(PLATFORM)"
+GENERATE[buildinf.h]=../util/mkbuildinf.pl "$(filter-out -I% -iremap% -fmacro-prefix-map%,$(CC) $(LIB_CFLAGS) $(CPPFLAGS_Q))" "$(PLATFORM)"
 DEPEND[buildinf.h]=../configdata.pm
 
 GENERATE[uplink-x86.s]=../ms/uplink-x86.pl $(PERLASM_SCHEME)
