--- miniupnpd/configure
+++ miniupnpd/configure
@@ -118,11 +118,11 @@
 	OS_VERSION=$(cat ./os.astlinux)
 fi
 
-# Tomato USB special case
+# FreshTomato special case
 if [ -f ../shared/tomato_version ]; then
 	OS_NAME=Tomato
-	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f2,3`
-	OS_VERSION="Tomato $TOMATO_VER"
+	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f1,2`
+	OS_VERSION="FreshTomato $TOMATO_VER"
 fi
 
 # OpenEmbedded special case
@@ -399,8 +399,8 @@
 		FW=iptables
 		;;
 	Tomato)
-		OS_NAME=UPnP
-		OS_URL=http://tomatousb.org/
+		OS_NAME=FreshTomato
+		OS_URL=https://freshtomato.org/
 		echo "" >> ${CONFIGFILE}
 		echo "#ifndef TOMATO" >> ${CONFIGFILE}
 		echo "#define TOMATO" >> ${CONFIGFILE}
@@ -412,6 +412,7 @@
 		echo "#ifdef TCONFIG_IPV6" >> ${CONFIGFILE}
 		echo "#define ENABLE_IPV6" >> ${CONFIGFILE}
 		echo "#endif" >> ${CONFIGFILE}
+		HAVE_IP_MREQN=1
 		FW=iptables
 		;;
 	Darwin)
@@ -495,8 +496,14 @@
 				fi
 				if [ "$IPTABLES_143" = "1" ] ; then
 					echo "LDLIBS += ${IPTABLESPATH}/libiptc/.libs/libip4tc.o" >> config.mk
+					if [ "$TCONFIG_IPV6" = "y" ]; then
+						echo "LDLIBS += ${IPTABLESPATH}/libiptc/.libs/libip6tc.o" >> config.mk
+					fi
 				else
 					echo "LDLIBS += ${IPTABLESPATH}/libiptc/libiptc.a" >> config.mk
+					if [ "$TCONFIG_IPV6" = "y" ]; then
+						echo "LDLIBS += ${IPTABLESPATH}/libiptc/libip6tc.a" >> config.mk
+					fi
 				fi
 			else
 				# OpenWrt
--- miniupnpd/Makefile.linux
+++ miniupnpd/Makefile.linux
@@ -29,11 +29,13 @@
 CFLAGS += -fno-common
 CFLAGS += -fstack-protector -fPIE
 CFLAGS += -D_FORTIFY_SOURCE=2
-CPPFLAGS += -D_GNU_SOURCE
+CFLAGS += -D_GNU_SOURCE
 CFLAGS += -Wall
 CFLAGS += -Wextra -Wstrict-prototypes -Wdeclaration-after-statement
 #CFLAGS += -Wno-missing-field-initializers
 #CFLAGS += -ansi	# iptables headers does use typeof which is a gcc extension
+CFLAGS += $(EXTRACFLAGS)
+CPPFLAGS += -I$(TOP)/shared
 LDFLAGS += -Wl,-z,now -Wl,-z,relro -pie
 CC ?= gcc
 RM = rm -f
@@ -41,7 +43,7 @@
 STRIP ?= strip
 PKG_CONFIG ?= pkg-config
 CP = cp
-DOXYGEN ?= doxygen
+#DOXYGEN ?= doxygen
 
 DEPFLAGS = -MM -MG -MT $(patsubst %.d,%.o,$@) -MT $@
 # -M : with system headers, -MM : without
@@ -52,7 +54,7 @@
 MANINSTALLDIR = $(INSTALLPREFIX)/share/man/man8
 
 include config.mk
-include $(SRCDIR)/gitrev.mk
+#include $(SRCDIR)/gitrev.mk
 include $(SRCDIR)/objects.mk
 
 # sources in netfilter/ directory
@@ -87,12 +89,12 @@
 ifeq ($(TARGET_OPENWRT),)
 LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l libssl)
 
-TEST := $(shell $(PKG_CONFIG) --exists uuid && echo 1)
-ifeq ($(TEST),1)
-LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l uuid)
-else
-$(info please install uuid-dev package / libuuid)
-endif # ($(TEST),1)
+#TEST := $(shell $(PKG_CONFIG) --exists uuid && echo 1)
+#ifeq ($(TEST),1)
+#LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l uuid)
+#else
+#$(info please install uuid-dev package / libuuid)
+#endif # ($(TEST),1)
 endif # ($(TARGET_OPENWRT,)
 
 ifneq ($(shell ldd --version | grep GLIBC),)
@@ -107,10 +109,11 @@
 
 TESTUPNPDESCGENOBJS = testupnpdescgen.o upnpdescgen.o
 
-EXECUTABLES = miniupnpd testupnpdescgen testgetifstats \
-              testupnppermissions miniupnpdctl testgetifaddr \
-              testgetroute testasyncsendto testportinuse \
-              testssdppktgen testminissdp
+EXECUTABLES = miniupnpd
+#EXECUTABLES = miniupnpd testupnpdescgen testgetifstats \
+#              testupnppermissions miniupnpdctl testgetifaddr \
+#              testgetroute testasyncsendto testportinuse \
+#              testssdppktgen testminissdp
 
 .PHONY:	all clean install depend dox
 
