--- miniupnpd/genconfig.sh	2018-05-21 23:16:44.294686204 +0200
+++ miniupnpd/genconfig.sh	2018-05-21 23:21:15.000000000 +0200
@@ -94,7 +94,7 @@ fi
 # Tomato USB special case
 if [ -f ../shared/tomato_version ]; then
 	OS_NAME=Tomato
-	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f2,3`
+	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f1,2`
 	OS_VERSION="Tomato $TOMATO_VER"
 fi
 
--- miniupnpd/Makefile.linux	2019-02-10 13:03:06.000000000 +0100
+++ miniupnpd/Makefile.linux	2019-02-24 16:48:42.102561446 +0100
@@ -27,7 +27,7 @@ CFLAGS += -fno-strict-aliasing
 CFLAGS += -fno-common
 CFLAGS += -fstack-protector -fPIE
 CFLAGS += -D_FORTIFY_SOURCE=2
-CPPFLAGS += -D_GNU_SOURCE
+CFLAGS += -D_GNU_SOURCE
 CFLAGS += -Wall
 CFLAGS += -Wextra -Wstrict-prototypes -Wdeclaration-after-statement
 #CFLAGS += -Wno-missing-field-initializers
@@ -40,6 +40,8 @@ STRIP ?= strip
 PKG_CONFIG ?= pkg-config
 CP = cp
 
+# for Tomato ARM (router/shared)
+CFLAGS += -I$(TOP)/shared
 
 INSTALLPREFIX ?= $(PREFIX)/usr
 SBININSTALLDIR = $(INSTALLPREFIX)/sbin
@@ -110,7 +112,13 @@ TEST := $(shell [ $(IPTVER1) -gt 1 ] ||
 ifeq ($(TEST), 1)
 CPPFLAGS := $(CPPFLAGS) -DIPTABLES_143
 # the following sucks, but works
+# for Tomato ARM - check IPv6 support
+ifeq ($(TCONFIG_IPV6),y)
 LDLIBS = $(IPTABLESPATH)/libiptc/.libs/libip4tc.o
+LDLIBS += $(IPTABLESPATH)/libiptc/.libs/libip6tc.o
+else # ($(TCONFIG_IPV6),y)
+LDLIBS = $(IPTABLESPATH)/libiptc/.libs/libip4tc.o
+endif # ($(TCONFIG_IPV6),y)
 #LDLIBS = $(IPTABLESPATH)/libiptc/.libs/libiptc.a
 else # ifeq ($(TEST), 1)
 LDLIBS = $(IPTABLESPATH)/libiptc/libiptc.a

--- miniupnpd/getifaddr.c	2018-07-06 14:03:21.000000000 +0200
+++ miniupnpd/getifaddr.c	2019-02-24 16:50:23.498396862 +0100
@@ -24,6 +24,8 @@
 #include "getifaddr.h"
 #if defined(USE_GETIFADDRS) || defined(ENABLE_IPV6) || defined(ENABLE_PCP)
 #include <ifaddrs.h>
+/* for Tomato ARM - Add getifaddr support to Tomato ARM builds (missing) */
+#include "ifaddrs.c"
 #endif
 
 int
