From 16ca9031d2ac197526eda3653793f4b908e35596 Mon Sep 17 00:00:00 2001
From: Fedor <fedork@ubuntu.(none)>
Date: Fri, 27 Aug 2010 14:56:06 -0400
Subject: [PATCH] pppd: populate gateway of default route with peer address
 (patch from OpenWRT) Ref: http://dev.openwrt.org/ticket/6259

---

--- pppd/pppd/sys-linux.c
+++ pppd/pppd/sys-linux.c
@@ -1642,6 +1642,9 @@ int sifdefaultroute (int unit, u_int32_t ouraddr, u_int32_t gateway)
     memset (&rt, 0, sizeof (rt));
     SET_SA_FAMILY (rt.rt_dst, AF_INET);
 
+    SET_SA_FAMILY(rt.rt_gateway, AF_INET);
+    SIN_ADDR(rt.rt_gateway) = gateway;
+
     rt.rt_dev = ifname;
 
     if (kernel_version > KVERSION(2,1,0)) {
@@ -1649,7 +1652,7 @@ int sifdefaultroute (int unit, u_int32_t ouraddr, u_int32_t gateway)
 	SIN_ADDR(rt.rt_genmask) = 0L;
     }
 
-    rt.rt_flags = RTF_UP;
+    rt.rt_flags = RTF_UP | RTF_GATEWAY;
     if (ioctl(sock_fd, SIOCADDRT, &rt) < 0) {
 	if ( ! ok_error ( errno ))
 	    error("default route ioctl(SIOCADDRT): %m");
-- 
2.10.5

