From a620fa767b3becd3b4e1eb8e070c6b18e602c8b5 Mon Sep 17 00:00:00 2001
From: Fedor <fedork@ubuntu.(none)>
Date: Sat, 9 Oct 2010 03:34:01 -0400
Subject: [PATCH] pppd: typo fix

---

--- pppd/pppd/ccp.c
+++ pppd/pppd/ccp.c
@@ -169,7 +169,7 @@ static option_t ccp_option_list[] = {
     { "+mppe-128", o_bool, &ccp_wantoptions[0].mppe_128,
       "require MPPE 128-bit encryption", 1, &ccp_allowoptions[0].mppe_128,
       OPT_ALIAS | OPT_PRIO },
-    { "nomppe-128", o_bool, &ccp_wantoptions[0].mppe_40,
+    { "nomppe-128", o_bool, &ccp_wantoptions[0].mppe_128,
       "don't allow MPPE 128-bit encryption", OPT_PRIOSUB | OPT_A2CLR,
       &ccp_allowoptions[0].mppe_128 },
     { "-mppe-128", o_bool, &ccp_wantoptions[0].mppe_128,
-- 
2.10.5

