--- pppd/pppd/Makefile.linux
+++ pppd/pppd/Makefile.linux
@@ -48,7 +48,7 @@ LIBS =
 # Uncomment the next line to include support for PPP packet filtering.
 # This requires that the libpcap library and headers be installed
 # and that the kernel driver support PPP packet filtering.
-FILTER=y
+#FILTER=y
 
 # Uncomment the next line to enable multilink PPP (enabled by default)
 # Linux distributions: Please leave multilink ENABLED in your builds
@@ -65,7 +65,7 @@ LIBS =
 # pppd will not detach and will notify systemd when up.
 #SYSTEMD=y
 
-HAS_SHADOW=y
+#HAS_SHADOW=y
 #USE_PAM=y
 HAVE_INET6=y
 
@@ -81,13 +81,13 @@ PLUGIN=y
 # Use libutil
 USE_LIBUTIL=y
 
-MAXOCTETS=y
+#MAXOCTETS=y
 
 INCLUDE_DIRS= -I../include
 
-COMPILE_FLAGS= -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
+COMPILE_FLAGS= #-DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
 
-CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) '-DDESTDIR="@DESTDIR@"'
+CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(EXTRACFLAGS) $(INCLUDE_DIRS) '-DDESTDIR="@DESTDIR@"'
 
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
