--- pppd/pppd/Makefile.linux
+++ pppd/pppd/Makefile.linux
@@ -3,6 +3,8 @@
 # Processed by configure to produce pppd/Makefile
 #
 
+include ../../common.mak
+
 CROSS_COMPILE=@CROSS_COMPILE@
 CC=$(CROSS_COMPILE)@CC@
 COPTS=@CFLAGS@
@@ -34,7 +36,7 @@
 include .depend
 endif
 
-LIBS = -lrt
+LIBS = $(if $(TCONFIG_BCMARM),-lrt)
 
 # Uncomment the next 2 lines to include support for Microsoft's
 # MS-CHAP authentication protocol.  Also, edit plugins/radius/Makefile.linux.
