--- pppd/pppd/plugins/pppol2tp/pppol2tp.c
+++ pppd/pppd/plugins/pppol2tp/pppol2tp.c
@@ -500,8 +500,10 @@
 	 */
 	add_notifier(&ip_up_notifier, pppol2tp_ip_up, NULL);
 	add_notifier(&ip_down_notifier, pppol2tp_ip_down, NULL);
+#ifdef INET6
 	add_notifier(&ipv6_up_notifier, pppol2tp_ip_up, NULL);
 	add_notifier(&ipv6_down_notifier, pppol2tp_ip_down, NULL);
+#endif
 }
 
 struct channel pppol2tp_channel = {
--- pppd/pppd/sys-linux.c
+++ pppd/pppd/sys-linux.c
@@ -220,7 +220,9 @@
 static unsigned char inbuf[512]; /* buffer for chars read from loopback */
 
 static int	if_is_up;	/* Interface has been marked up */
+#ifdef INET6
 static int	if6_is_up;	/* Interface has been marked up for IPv6, to help differentiate */
+#endif
 static int	have_default_route;	/* Gateway for default route added */
 static int	have_default_route6;	/* Gateway for default IPv6 route added */
 static struct	rtentry old_def_rt;	/* Old default route */
@@ -361,8 +363,10 @@
 	if_is_up = 0;
 	sifdown(0);
     }
+#ifdef INET6
     if (if6_is_up)
 	sif6down(0);
+#endif
 
 /*
  * Delete any routes through the device.
