--- util-linux/libuuid/src/unparse.c
+++ util-linux/libuuid/src/unparse.c
@@ -33,6 +33,7 @@
  */
 
 #include <stdio.h>
+#include "c.h"
 
 #include "uuidP.h"
 
--- util-linux/libuuid/src/uuid.h
+++ util-linux/libuuid/src/uuid.h
@@ -32,8 +32,8 @@
  * %End-Header%
  */
 
-#ifndef _UUID_UUID_H
-#define _UUID_UUID_H
+#ifndef _UL_LIBUUID_UUID_H
+#define _UL_LIBUUID_UUID_H
 
 #include <sys/types.h>
 #ifndef _WIN32
@@ -120,4 +120,4 @@
 }
 #endif
 
-#endif /* _UUID_UUID_H */
+#endif /* _UL_LIBUUID_UUID_H */
