--- ffmpeg/libavcodec/h264.c
+++ ffmpeg/libavcodec/h264.c
@@ -1726,8 +1726,11 @@
                                     idct_dc_add(ptr, h->mb + (i*16+p*256 << pixel_shift), linesize);
                                 else
                                     idct_add   (ptr, h->mb + (i*16+p*256 << pixel_shift), linesize);
-                            }else
+                            }
+#if CONFIG_SVQ3_DECODER
+                            else
                                 ff_svq3_add_idct_c(ptr, h->mb + i*16+p*256, linesize, qscale, 0);
+#endif
                         }
                     }
                 }
@@ -1746,8 +1749,11 @@
                         dctcoef_set(h->mb+p*256, pixel_shift, dc_mapping[i], dctcoef_get(h->mb_luma_dc[p], pixel_shift, i));
                 }
             }
-        }else
+        }
+#if CONFIG_SVQ3_DECODER
+        else
             ff_svq3_luma_dc_dequant_idct_c(h->mb+p*256, h->mb_luma_dc[p], qscale);
+#endif
     }
 }
 
@@ -1790,7 +1796,9 @@
                     }
                 }
             }
-        }else{
+        }
+#if CONFIG_SVQ3_DECODER
+        else{
             for(i=0; i<16; i++){
                 if(h->non_zero_count_cache[ scan8[i+p*16] ] || h->mb[i*16+p*256]){ //FIXME benchmark weird rule, & below
                     uint8_t * const ptr= dest_y + block_offset[i];
@@ -1798,6 +1806,7 @@
                 }
             }
         }
+#endif
     }
 }
 
