--- busybox/networking/udhcp/dhcpc.c
+++ busybox/networking/udhcp/dhcpc.c
@@ -716,6 +716,7 @@
 static NOINLINE int send_discover(uint32_t xid, uint32_t requested)
 {
 	struct dhcp_packet packet;
+	static int msgs = 0;
 
 	/* Fill in: op, htype, hlen, cookie, chaddr fields,
 	 * random xid field (we override it below),
@@ -733,7 +734,8 @@
 	 */
 	add_client_options(&packet);
 
-	bb_info_msg("sending %s", "discover");
+	if (msgs++ < 3)
+		bb_info_msg("sending %s", "discover");
 	return raw_bcast_from_client_data_ifindex(&packet, INADDR_ANY);
 }
 
