--- busybox/libbb/udp_io.c
+++ busybox/libbb/udp_io.c
@@ -19,7 +19,14 @@ socket_want_pktinfo(int fd UNUSED_PARAM)
 	setsockopt_1(fd, IPPROTO_IP, IP_PKTINFO);
 #endif
 #if ENABLE_FEATURE_IPV6 && defined(IPV6_PKTINFO)
+# ifdef IPV6_RECVPKTINFO
+	setsockopt_1(fd, SOL_IPV6, IPV6_RECVPKTINFO);
+# ifdef IPV6_2292PKTINFO
+	setsockopt_1(fd, SOL_IPV6, IPV6_2292PKTINFO);
+# endif
+# else
 	setsockopt_1(fd, IPPROTO_IPV6, IPV6_PKTINFO);
+# endif
 #endif
 }
 
@@ -168,14 +175,19 @@ recv_from_to(int fd, void *buf, size_t l
 		}
 # if ENABLE_FEATURE_IPV6 && defined(IPV6_PKTINFO)
 		if (cmsgptr->cmsg_level == IPPROTO_IPV6
-		 && cmsgptr->cmsg_type == IPV6_PKTINFO
-		) {
+		 && (cmsgptr->cmsg_type == IPV6_PKTINFO
+# if defined(IPV6_RECVPKTINFO) && defined(IPV6_2292PKTINFO)
+		  || cmsgptr->cmsg_type == IPV6_2292PKTINFO
+# endif
+		)) {
 			const int IPI6_ADDR_OFF = offsetof(struct in6_pktinfo, ipi6_addr);
+			const int IPI6_IFINDEX_OFF = offsetof(struct in6_pktinfo, ipi6_ifindex);
 			to->sa_family = AF_INET6;
 			/*#  define pktinfo(cmsgptr) ( (struct in6_pktinfo*)(CMSG_DATA(cmsgptr)) )*/
 			/*to6->sin6_addr = pktinfo(cmsgptr)->ipi6_addr; - may be unaligned */
 			memcpy(&to6->sin6_addr, (char*)(CMSG_DATA(cmsgptr)) + IPI6_ADDR_OFF, sizeof(to6->sin6_addr));
 			/*to6->sin6_port = 123; */
+			move_from_unaligned_int(to6->sin6_scope_id, (char*)(CMSG_DATA(cmsgptr)) + IPI6_IFINDEX_OFF);
 			break;
 		}
 # endif
--- busybox/networking/traceroute.c
+++ busybox/networking/traceroute.c
@@ -903,12 +903,7 @@ common_traceroute_main(int op, char **ar
 #if ENABLE_TRACEROUTE6
 	if (af == AF_INET6) {
 		xmove_fd(xsocket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6), rcvsock);
-# ifdef IPV6_RECVPKTINFO
-		setsockopt_1(rcvsock, SOL_IPV6, IPV6_RECVPKTINFO);
-		setsockopt_1(rcvsock, SOL_IPV6, IPV6_2292PKTINFO);
-# else
-		setsockopt_1(rcvsock, SOL_IPV6, IPV6_PKTINFO);
-# endif
+		socket_want_pktinfo(rcvsock);
 	} else
 #endif
 	{
