--- hotplug2/hotplug2_utils.c
+++ hotplug2/hotplug2_utils.c
@@ -19,7 +19,11 @@
  *
  * Returns: Seqnum as read from sysfs
  */
+#if !defined(CONFIG_TOOLCHAIN53) && !defined(CONFIG_TOOLCHAIN73)
 inline event_seqnum_t get_kernel_seqnum() {
+#else
+extern inline event_seqnum_t get_kernel_seqnum() {
+#endif
 	FILE *fp;
 	
 	char filename[64];
@@ -44,7 +48,11 @@
  *
  * Returns: Socket fd if succesful, -1 otherwise.
  */
+#if !defined(CONFIG_TOOLCHAIN53) && !defined(CONFIG_TOOLCHAIN73)
 inline int init_netlink_socket(int type) {
+#else
+extern inline int init_netlink_socket(int type) {
+#endif
 	int netlink_socket;
 	struct sockaddr_nl snl;
 	int buffersize = 16 * 1024 * 1024;
--- hotplug2/mem_utils.c
+++ hotplug2/mem_utils.c
@@ -16,7 +16,11 @@
  *
  * Returns: Pointer to freshly allocated memory
  */
+#if !defined(CONFIG_TOOLCHAIN53) && !defined(CONFIG_TOOLCHAIN73)
 inline void *xmalloc(size_t size) {
+#else
+extern inline void *xmalloc(size_t size) {
+#endif
 	void *ptr;
 	ptr = malloc(size);
 	if (ptr == NULL) {
@@ -34,7 +38,11 @@
  *
  * Returns: Pointer to reallocated memory
  */
+#if !defined(CONFIG_TOOLCHAIN53) && !defined(CONFIG_TOOLCHAIN73)
 inline void *xrealloc(void *inptr, size_t size) {
+#else
+extern inline void *xrealloc(void *inptr, size_t size) {
+#endif
 	void *ptr;
 	ptr = realloc(inptr, size);
 	if (ptr == NULL) {
--- hotplug2/hotplug2.c
+++ hotplug2/hotplug2.c
@@ -62,7 +62,11 @@
  *
  * Returns: void
  */
+#ifndef CONFIG_TOOLCHAIN73
 inline void free_hotplug2_event(struct hotplug2_event_t *event) {
+#else
+void free_hotplug2_event(struct hotplug2_event_t *event) {
+#endif
 	int i;
 	
 	for (i = 0; i < event->env_vars_c; i++) {
@@ -81,7 +85,11 @@
  *
  * Returns: Macro of the given action
  */
+#ifndef CONFIG_TOOLCHAIN73
 inline int get_hotplug2_event_action(char *action) {
+#else
+int get_hotplug2_event_action(char *action) {
+#endif
 	if (!strcmp(action, "add"))
 		return ACTION_ADD;
 	
@@ -161,7 +169,11 @@
  *
  * Returns: A copy of the source event structure
  */
+#ifndef CONFIG_TOOLCHAIN73
 inline struct hotplug2_event_t *dup_hotplug2_event(struct hotplug2_event_t *src) {
+#else
+struct hotplug2_event_t *dup_hotplug2_event(struct hotplug2_event_t *src) {
+#endif
 	struct hotplug2_event_t *dest;
 	int i;
 	
@@ -189,7 +201,11 @@
  *
  * Returns: A new event structure
  */
+#ifndef CONFIG_TOOLCHAIN73
 inline struct hotplug2_event_t *get_hotplug2_event(char *event_str, int size) {
+#else
+struct hotplug2_event_t *get_hotplug2_event(char *event_str, int size) {
+#endif
 	char *ptr;
 	struct hotplug2_event_t *event;
 	int skip;
