--- dnscrypt/configure.ac
+++ dnscrypt/configure.ac
@@ -246,12 +246,12 @@
 
 AC_SUBST([PROBES_SOURCE])
 
-AS_IF([test -d /usr/local/include], [
-  CPPFLAGS="$CPPFLAGS -I/usr/local/include"
-])
-AS_IF([test -d /usr/local/lib], [
-  LDFLAGS="$LDFLAGS -L/usr/local/lib"
-])
+dnl AS_IF([test -d /usr/local/include], [
+dnl  CPPFLAGS="$CPPFLAGS -I/usr/local/include"
+dnl ])
+dnl AS_IF([test -d /usr/local/lib], [
+dnl   LDFLAGS="$LDFLAGS -L/usr/local/lib"
+dnl ])
 
 AS_IF([test -d /usr/local/opt/libressl/include], [
   CPPFLAGS="$CPPFLAGS -I/usr/local/opt/libressl/include"
