--- portmap/Makefile
+++ portmap/Makefile
@@ -124,10 +124,10 @@ CPPFLAGS += -DIGNORE_SIGCHLD	# AIX 4.x,
 
 all:	portmap pmap_dump pmap_set portmap.man
 
-CPPFLAGS += $(HOSTS_ACCESS)
-portmap: CFLAGS   += -fpie
-portmap: LDLIBS   += $(WRAP_LIB)
-portmap: LDFLAGS  += -pie
+#CPPFLAGS += $(HOSTS_ACCESS)
+#portmap: CFLAGS   += -fpie
+#portmap: LDLIBS   += $(WRAP_LIB)
+#portmap: LDFLAGS  += -pie
 portmap: portmap.o pmap_check.o from_local.o
 
 from_local: CPPFLAGS += -DTEST
--- portmap/pmap_check.c
+++ portmap/pmap_check.c
@@ -44,7 +44,7 @@
 #include <netinet/in.h>
 #include <rpc/rpcent.h>
 #endif
-#include <tcpd.h>
+//#include <tcpd.h>
 #include <arpa/inet.h>
 #include <grp.h>
 
