--- libjson-c/CMakeLists.txt
+++ libjson-c/CMakeLists.txt
@@ -48,9 +48,9 @@
 add_subdirectory(tests)
 endif()
 
-if (NOT MSVC)  # cmd line apps don't built on Windows currently.
-add_subdirectory(apps)
-endif()
+#if (NOT MSVC)  # cmd line apps don't built on Windows currently.
+#add_subdirectory(apps)
+#endif()
 
 # Set some packaging variables.
 set(CPACK_PACKAGE_NAME              "${PROJECT_NAME}")
@@ -433,7 +433,7 @@
 include_directories(${PROJECT_SOURCE_DIR})
 include_directories(${PROJECT_BINARY_DIR})
 
-add_subdirectory(doc)
+#add_subdirectory(doc)
 
 # "uninstall" custom target for make generators in unix like operating systems
 # and if that target is not present
