--- libcurl/m4/curl-functions.m4
+++ libcurl/m4/curl-functions.m4
@@ -5,7 +5,7 @@
 #                            | (__| |_| |  _ <| |___
 #                             \___|\___/|_| \_\_____|
 #
-# Copyright (C) 1998 - 2021, Daniel Stenberg, <daniel@haxx.se>, et al.
+# Copyright (C) 1998 - 2022, Daniel Stenberg, <daniel@haxx.se>, et al.
 #
 # This software is licensed as described in the file COPYING, which
 # you should have received as part of this distribution. The terms
@@ -1149,7 +1149,7 @@
   tst_allow_fcntl_o_nonblock="unknown"
   #
   case $host_os in
-    sunos4* | aix3* | beos*)
+    sunos4* | aix3*)
       dnl O_NONBLOCK does not work on these platforms
       curl_disallow_fcntl_o_nonblock="yes"
       ;;
@@ -6515,16 +6515,17 @@
 
 AC_DEFUN([CURL_RUN_IFELSE], [
    case $host_os in
-     darwin*) library_path_var=DYLD_LIBRARY_PATH ;;
-     *)       library_path_var=LD_LIBRARY_PATH ;;
+     darwin*)
+      AC_RUN_IFELSE([AC_LANG_SOURCE([$1])], $2, $3, $4)
+     ;;
+     *)
+      old=$LD_LIBRARY_PATH
+      LD_LIBRARY_PATH=$CURL_LIBRARY_PATH:$old
+      export LD_LIBRARY_PATH
+      AC_RUN_IFELSE([AC_LANG_SOURCE([$1])], $2, $3, $4)
+      LD_LIBRARY_PATH=$old # restore
+     ;;
    esac
-
-   eval "old=$$library_path_var"
-   eval "$library_path_var=\$CURL_LIBRARY_PATH:\$old"
-
-   eval "export $library_path_var"
-   AC_RUN_IFELSE([AC_LANG_SOURCE([$1])], $2, $3, $4)
-   eval "$library_path_var=\$old" # restore
 ])
 
 dnl CURL_COVERAGE
@@ -6561,7 +6562,7 @@
     fi
 
     CPPFLAGS="$CPPFLAGS -DNDEBUG"
-    CFLAGS="$CLAGS -O0 -g -fprofile-arcs -ftest-coverage"
+    CFLAGS="$CFLAGS -O0 -g -fprofile-arcs -ftest-coverage"
     LIBS="$LIBS -lgcov"
   fi
 ])
