--- wolfssl/wolfssl/wolfcrypt/wc_port.h
+++ wolfssl/wolfssl/wolfcrypt/wc_port.h
@@ -1245,7 +1245,7 @@
 #ifndef WOLFSSL_NO_FENCE
     #if defined (__i386__) || defined(__x86_64__)
         #define XFENCE() XASM_VOLATILE("lfence")
-    #elif (defined (__arm__) && (__ARM_ARCH > 6)) || defined(__aarch64__)
+    #elif (defined (__arm__) && defined(__ARM_ARCH) && (__ARM_ARCH > 6)) || defined(__aarch64__)
         #define XFENCE() XASM_VOLATILE("isb")
     #elif defined(__riscv)
         #define XFENCE() XASM_VOLATILE("fence")
