--- wsdd2/wsdd2.c
+++ wsdd2/wsdd2.c
@@ -701,6 +701,7 @@
 	}
 
 	struct ifaddrs *ifaddrs;
+	struct in6_addr in6addr;
 	fd_set fds;
 	int svn, rv = 0, nfds = -1;
 	struct endpoint *ep, *badep = NULL;
@@ -752,6 +753,13 @@
 						continue;
 				}
 
+				/* only process ipv6 linklocal */
+				if (ifa->ifa_addr->sa_family == AF_INET6) {
+					memcpy(&in6addr, &((struct sockaddr_in6 *)ifa->ifa_addr)->sin6_addr, sizeof(in6addr));
+					if (!IN6_IS_ADDR_LINKLOCAL(&in6addr) && !IN6_IS_ADDR_MC_LINKLOCAL(&in6addr))
+						continue;
+				}
+
 				char ifaddr[_ADDRSTRLEN];
 				void *addr =
 					_SIN_ADDR((_saddr_t *)ifa->ifa_addr);
