--- getdns/stubby/src/stubby.c
+++ getdns/stubby/src/stubby.c
@@ -192,6 +192,8 @@ print_usage(FILE *out)
 	fprintf(out, "\t\t\t\"%s\"\n", system_conf_fn);
 	fprintf(out, "\t\tA default file (Using Strict mode) is installed as\n");
 	fprintf(out, "\t\t\t\"%s\"\n", system_conf_fn);
+	fprintf(out, "\t-F\t<filename>\n");
+	fprintf(out, "\t\tWrite log output to <filename> (used with -g background mode)\n");
 #if !defined(STUBBY_ON_WINDOWS) && !defined(GETDNS_ON_WINDOWS)
 	fprintf(out, "\t-g\tRun stubby in background (default is foreground)\n");
 #endif
@@ -759,6 +761,8 @@ main(int argc, char **argv)
 {
 	char *conf_fn;
 	const char *custom_config_fn = NULL;
+	const char *custom_log_fn = NULL;
+	FILE *lfn;
 	int found_conf = 0;
 	int print_api_info = 0;
 	int log_connections = 0;
@@ -770,11 +774,22 @@ main(int argc, char **argv)
 	getdns_list *api_info_keys = NULL;
 	getdns_bindata *api_info_key = NULL;
 
-	while ((opt = getopt(argc, argv, "C:ighlv:V")) != -1) {
+	while ((opt = getopt(argc, argv, "C:F:ighlv:V")) != -1) {
 		switch (opt) {
 		case 'C':
 			custom_config_fn = optarg;
 			break;
+		case 'F':
+			custom_log_fn = optarg;
+			if ((lfn = fopen(custom_log_fn, "w")) != NULL) {
+				dup2(fileno(lfn), STDOUT_FILENO);
+				dup2(fileno(lfn), STDERR_FILENO);
+				fclose(lfn);
+			} else {
+				fprintf(stderr, "Could not open log file '%s'\n", custom_log_fn);
+				exit(EXIT_FAILURE);
+			}
+			break;
 		case 'g':
 			run_in_foreground = 0;
 			break;
@@ -798,7 +813,7 @@ main(int argc, char **argv)
 				exit(EXIT_FAILURE);
 			}
 			break;
-                case 'V':
+		case 'V':
 			print_version(stdout);
 			exit(EXIT_SUCCESS);
 		default:
