--- getdns/cmake/include/cmakeconfig.h.in
+++ getdns/cmake/include/cmakeconfig.h.in
@@ -103,6 +103,8 @@
 #cmakedefine HAVE_PTHREAD               1
 #cmakedefine HAVE_WINDOWS_THREADS       1
 
+#cmakedefine HAVE_ATTR_UNUSED   1
+
 #cmakedefine RUNSTATEDIR                "@RUNSTATEDIR@"
 #cmakedefine TRUST_ANCHOR_FILE          "@PATH_TRUST_ANCHOR_FILE@"
 #cmakedefine GETDNS_FN_RESOLVCONF       "@PATH_RESOLVCONF@"
--- getdns/cmake/tests/test_unused_attr.c
+++ getdns/cmake/tests/test_unused_attr.c
@@ -0,0 +1,13 @@
+#include <stdio.h>
+
+void f (char *u __attribute__((unused)));
+
+void f(char *u)
+{
+}
+
+int main (int ac, char *av[])
+{
+    f("str");
+    return 0;
+}
--- getdns/CMakeLists.txt
+++ getdns/CMakeLists.txt
@@ -1006,6 +1006,12 @@ if (BUILD_TESTING)
   endif()
 endif ()
 
+# Does the compiler accept the "unused" attribute?
+try_compile(HAVE_ATTR_UNUSED
+  ${CMAKE_CURRENT_BINARY_DIR}
+  ${CMAKE_CURRENT_SOURCE_DIR}/cmake/tests/test_unused_attr.c
+  )
+
 # Substitutions in files.
 configure_file(cmake/include/cmakeconfig.h.in config.h)
 configure_file(src/getdns/getdns.h.in getdns/getdns.h)
