--- libubox/uloop.h
+++ libubox/uloop.h
@@ -94,7 +94,11 @@
 int uloop_timeout_add(struct uloop_timeout *timeout);
 int uloop_timeout_set(struct uloop_timeout *timeout, int msecs);
 int uloop_timeout_cancel(struct uloop_timeout *timeout);
+#ifdef TCONFIG_BCMARM
 int uloop_timeout_remaining(struct uloop_timeout *timeout) __attribute__((deprecated("use uloop_timeout_remaining64")));
+#else
+int uloop_timeout_remaining(struct uloop_timeout *timeout) __attribute__((deprecated));
+#endif
 int64_t uloop_timeout_remaining64(struct uloop_timeout *timeout);
 
 int uloop_process_add(struct uloop_process *p);
