--- tinc/configure.ac
+++ tinc/configure.ac
@@ -3,9 +3,9 @@ dnl Process this file with autoconf to p
 origcflags="$CFLAGS"
 
 AC_PREREQ(2.69)
-AC_INIT([tinc], m4_esyscmd_s((git describe || echo UNKNOWN) | sed 's/release-//'))
+AC_INIT([tinc], [1.1pre18])
 AC_CONFIG_SRCDIR([src/tincd.c])
-AM_INIT_AUTOMAKE([std-options subdir-objects nostdinc silent-rules -Wall])
+AM_INIT_AUTOMAKE([check-news std-options subdir-objects nostdinc silent-rules -Wall])
 AC_CONFIG_HEADERS([config.h])
 AC_CONFIG_MACRO_DIR([m4])
 AM_SILENT_RULES([yes])
--- tinc/src/version.c
+++ tinc/src/version.c
@@ -24,6 +24,7 @@
 /* This file is always rebuilt (even if there are no changes) so that the following is updated */
 const char *const BUILD_DATE = __DATE__;
 const char *const BUILD_TIME = __TIME__;
+#undef GIT_DESCRIPTION
 #ifdef GIT_DESCRIPTION
 const char *const BUILD_VERSION = GIT_DESCRIPTION;
 #else
