--- libusb10/libusb/os/linux_netlink.c
+++ libusb10/libusb/os/linux_netlink.c
@@ -34,8 +34,8 @@
 #ifdef HAVE_ASM_TYPES_H
 #include <asm/types.h>
 #endif
-#include <linux/netlink.h>
 #include <sys/socket.h>
+#include <linux/netlink.h>
 
 #define NL_GROUP_KERNEL 1
 
--- libusb10/configure.ac
+++ libusb10/configure.ac
@@ -136,8 +136,12 @@
 		AC_CHECK_LIB([udev], [udev_new], [], [AC_MSG_ERROR([udev support requested but libudev not installed])])
 	else
 		AC_CHECK_HEADERS([asm/types.h])
-		AC_CHECK_HEADER([linux/netlink.h], [], [AC_MSG_ERROR([Linux netlink header not found])])
-		AC_CHECK_HEADER([sys/socket.h], [], [AC_MSG_ERROR([Linux socket header not found])])
+		AC_CHECK_HEADERS([sys/socket.h linux/netlink.h], [], [AC_MSG_ERROR([Linux netlink headers not found])], [
+#ifdef HAVE_ASM_TYPES_H
+#include <asm/types.h>
+#endif
+#include <sys/socket.h>
+])
 	fi
 	if test "x$is_backend_android" != xyes; then
 		THREAD_CFLAGS="-pthread"
