--- php/ext/libxml/php_libxml.h
+++ php/ext/libxml/php_libxml.h
@@ -155,7 +155,7 @@
 	_Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
 # define PHP_LIBXML_IGNORE_DEPRECATIONS_END \
 	_Pragma("clang diagnostic pop")
-#elif defined(__GNUC__)
+#elif defined(__GNUC__) && __GNUC__ >= 5
 # define PHP_LIBXML_IGNORE_DEPRECATIONS_START \
 	_Pragma("GCC diagnostic push") \
 	_Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
