--- mysql/extra/Makefile.am
+++ mysql/extra/Makefile.am
@@ -32,7 +32,7 @@ DIST_SUBDIRS =		yassl
 $(top_builddir)/include/mysqld_error.h: comp_err.c \
 			$(top_srcdir)/sql/share/errmsg.txt
 	$(MAKE) $(AM_MAKEFLAGS) comp_err$(EXEEXT)
-	$(top_builddir)/extra/comp_err$(EXEEXT) \
+	$(top_builddir)/host/comp_err$(EXEEXT) \
 	--charset=$(top_srcdir)/sql/share/charsets \
 	--out-dir=$(top_builddir)/sql/share/ \
 	--header_file=$(top_builddir)/include/mysqld_error.h \
--- mysql/extra/Makefile.in
+++ mysql/extra/Makefile.in
@@ -938,7 +938,7 @@ uninstall-am: uninstall-binPROGRAMS unin
 $(top_builddir)/include/mysqld_error.h: comp_err.c \
 			$(top_srcdir)/sql/share/errmsg.txt
 	$(MAKE) $(AM_MAKEFLAGS) comp_err$(EXEEXT)
-	$(top_builddir)/extra/comp_err$(EXEEXT) \
+	$(top_builddir)/host/comp_err$(EXEEXT) \
 	--charset=$(top_srcdir)/sql/share/charsets \
 	--out-dir=$(top_builddir)/sql/share/ \
 	--header_file=$(top_builddir)/include/mysqld_error.h \
--- mysql/include/my_global.h
+++ mysql/include/my_global.h
@@ -894,6 +894,7 @@ typedef SOCKET_SIZE_TYPE size_socket;
 #endif /* isfinite */
 
 #ifndef HAVE_ISNAN
+#undef isnan //bwq518
 #define isnan(x) ((x) != (x))
 #endif
 
--- mysql/scripts/Makefile.am
+++ mysql/scripts/Makefile.am
@@ -121,7 +121,7 @@ mysql_fix_privilege_tables.sql: mysql_sy
 mysql_fix_privilege_tables_sql.c: comp_sql.c mysql_fix_privilege_tables.sql
 	$(MAKE) $(AM_MAKEFLAGS) comp_sql$(EXEEXT)
 	sleep 2
-	$(top_builddir)/scripts/comp_sql$(EXEEXT) \
+	$(top_builddir)/host/comp_sql$(EXEEXT) \
 	  mysql_fix_privilege_tables \
 	    $(top_srcdir)/scripts/mysql_fix_privilege_tables.sql $@
 
--- mysql/scripts/Makefile.in
+++ mysql/scripts/Makefile.in
@@ -804,7 +804,7 @@ mysql_fix_privilege_tables.sql: mysql_sy
 mysql_fix_privilege_tables_sql.c: comp_sql.c mysql_fix_privilege_tables.sql
 	$(MAKE) $(AM_MAKEFLAGS) comp_sql$(EXEEXT)
 	sleep 2
-	$(top_builddir)/scripts/comp_sql$(EXEEXT) \
+	$(top_builddir)/host/comp_sql$(EXEEXT) \
 	  mysql_fix_privilege_tables \
 	    $(top_srcdir)/scripts/mysql_fix_privilege_tables.sql $@
 
--- mysql/sql/Makefile.am
+++ mysql/sql/Makefile.am
@@ -177,7 +177,7 @@ link_sources:
 # this avoid the rebuild of the built files in a source dist
 lex_hash.h:	gen_lex_hash.cc lex.h
 		$(MAKE) $(AM_MAKEFLAGS) gen_lex_hash$(EXEEXT)
-		./gen_lex_hash$(EXEEXT) > $@-t
+		../host/gen_lex_hash$(EXEEXT) > $@-t
 		$(MV) $@-t $@
 
 # For testing of udf_example.so
--- mysql/sql/Makefile.in
+++ mysql/sql/Makefile.in
@@ -1310,7 +1310,7 @@ link_sources:
 # this avoid the rebuild of the built files in a source dist
 lex_hash.h:	gen_lex_hash.cc lex.h
 		$(MAKE) $(AM_MAKEFLAGS) gen_lex_hash$(EXEEXT)
-		./gen_lex_hash$(EXEEXT) > $@-t
+		../host/gen_lex_hash$(EXEEXT) > $@-t
 		$(MV) $@-t $@
 
 # We might have some stuff not built in this build, but that we want to install
